akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala

```diff
@@ -85,6 +86,32 @@ trait FileUploadDirectives {
       case Some(tuple) ⇒ provide(tuple)
       case None        ⇒ reject(MissingFormFieldRejection(fieldName))
     }
+
+  /**
+   * Collects each body part that is a multipart file as a tuple containing metadata and a `Source`
+   * for streaming the file contents somewhere. If there is no such field the request will be rejected.
+   *
+   * @group fileupload
+   */
+  def fileMultiUpload(fieldName: String): Directive1[immutable.Seq[(FileInfo, Source[ByteString, Any])]] =
+    entity(as[Multipart.FormData]).flatMap { formData ⇒
+      extractRequestContext.flatMap { ctx ⇒
+        implicit val mat = ctx.materializer
+        implicit val ec = ctx.executionContext
+
+        val partSource: Source[(FileInfo, Source[ByteString, Any]), Any] = formData.parts
+          .filter(part ⇒ part.filename.isDefined && part.name == fieldName)
+          .map(part ⇒ (FileInfo(part.name, part.filename.get, part.entity.contentType), part.entity.dataBytes))
+
+        val partF = partSource.runWith(Sink.seq[(FileInfo, Source[ByteString, Any])])
```

Comment:
Unfortunately, this approach will not work if the file uploads are big enough. You try to collect ongoing sources for the parts into a seq. However, without reading the first source you will not get the second source, so that the `seq` will never complete and the stream will deadlock. This is an unfortunate consequence of the squential `multipart` format. It will work with smaller payloads that fit into all the buffers but if the whole request doesn't fit into the buffers it will deadlock.

I think we need to do the same that other web servers / frameworks are doing and buffer the incoming files on the disk so that we can drain the whole request body before giving out any data.